File: lib/models/question_id.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';

@immutable
class QuestionId {
  const QuestionId({
    required this.questionId,
    required this.authId,
    required this.creAt,
  });

  final String questionId;
  final String authId;
  final DateTime creAt;

  static Map<String, dynamic> forSet(String authId, String questionId) {
    return <String, dynamic>{
      'questionId': questionId,
      'authId': authId,
      'creAt': FieldValue.serverTimestamp(),
    };
  }

  // fromQuestion
  factory QuestionId.fromQuestion({
    required String questionId,
    required String authId,
  }) {
    return QuestionId(
      questionId: questionId,
      authId: authId,
      creAt: DateTime.now(),
    );
  }

  // fromFireatore
  factory QuestionId.fromFirestore(Map<String, dynamic> data) {
    return QuestionId(
      questionId: data['questionId'] as String? ?? '',
      authId: data['authId'] as String? ?? '',
      creAt: (data['creAt'] as Timestamp?)?.toDate() ?? DateTime.now(),
    );
  }

  // hashCode
  @override
  int get hashCode => questionId.hashCode ^ authId.hashCode ^ creAt.hashCode;

  // operator ==
  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other.runtimeType != runtimeType) return false;
    final QuestionId otherQuestionId = other as QuestionId;
    return questionId == otherQuestionId.questionId &&
        authId == otherQuestionId.authId &&
        creAt == otherQuestionId.creAt;
  }
}

File: functions/src/models/question_id.ts
import { Timestamp } from "firebase-admin/firestore";

/**
 * 質問IDを表すクラス
 */
export class QuestionId {
  readonly authId: string;
  readonly questionId: string;
  readonly creAt: Timestamp;

  /**
   * QuestionIdのインスタンスを作成します。
   * @param {string} authId - 認証ID
   * @param {string} questionId - 質問ID
   * @param {Timestamp} creAt - 作成日時
   */
  constructor(authId: string, questionId: string, creAt: Timestamp) {
    this.authId = authId;
    this.questionId = questionId;
    this.creAt = creAt;
  }

  /**
   * FirebaseFirestoreに保存するためのデータを作成します。
   * @param {string} authId - 認証ID
   * @param {string} questionId - 質問ID
   * @return {FirebaseFirestore.DocumentData} - 保存するデータ
   */
  static forSet(
    authId: string,
    questionId: string
  ): FirebaseFirestore.DocumentData {
    return {
      authId: authId,
      questionId: questionId,
      creAt: Timestamp.now(),
    };
  }
}
